File: src/client/plan.py

import loomplan_pb2


POLICY_STANDARD = loomplan_pb2.Task.POLICY_STANDARD
POLICY_SIMPLE = loomplan_pb2.Task.POLICY_SIMPLE
POLICY_SCHEDULER = loomplan_pb2.Task.POLICY_SCHEDULER


class Task(object):

    task_type = None
    inputs = ()
    id = None
    config = ""
    policy = POLICY_STANDARD
    resource_request = None
    label = None

    def set_message(self, msg, symbols, requests):
        msg.config = self.config
        msg.task_type = symbols[self.task_type]
        msg.input_ids.extend(t.id for t in self.inputs)
        msg.policy = self.policy
        if self.resource_request:
            msg.resource_request_index = requests.index(self.resource_request)
        if self.label:
            msg.label = self.label


class ResourceRequest(object):

    def __init__(self):
        self.resources = {}

    def add_resource(self, name, value):
        self.resources[name] = value

    @property
    def names(self):
        return self.resources.keys()

    def set_message(self, msg, symbols):
        for name, value in self.resources.items():
            r = msg.resources.add()
            r.resource_type = symbols[name]
            r.value = value

    def __eq__(self, other):
        if not isinstance(other, ResourceRequest):
            return False
        return self.resources == other.resources

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return hash(tuple(sorted(self.resources.items())))


class Plan(object):

    def __init__(self):
        self.tasks = []

    def add(self, task):
        assert task.id is None
        task.id = len(self.tasks)
        self.tasks.append(task)
        return task

    def collect_symbols(self):
        symbols = set()
        for task in self.tasks:
            if task.resource_request:
                symbols.update(task.resource_request.names)
            symbols.add(task.task_type)
        return symbols

    def set_message(self, msg, symbols):
        requests = set()
        for task in self.tasks:
            if task.resource_request:
                requests.add(task.resource_request)
        requests = list(requests)

        for request in requests:
            r = msg.resource_requests.add()
            request.set_message(r, symbols)

        for task in self.tasks:
            t = msg.tasks.add()
            task.set_message(t, symbols, requests)
        return msg

File: src/server/plan.cpp
#include "plan.h"

#include <algorithm>

#include "libloom/loomplan.pb.h"
#include "libloom/log.h"

static PlanNode::Policy read_task_policy(loomplan::Task_Policy policy) {
    switch(policy) {
        case loomplan::Task_Policy_POLICY_STANDARD:
            return PlanNode::POLICY_STANDARD;
        case loomplan::Task_Policy_POLICY_SIMPLE:
            return PlanNode::POLICY_SIMPLE;
        case loomplan::Task_Policy_POLICY_SCHEDULER:
            return PlanNode::POLICY_SCHEDULER;
        default:
            loom::llog->critical("Invalid task policy");
            exit(1);
    }
}

Plan::Plan()
{

}

Plan::Plan(const loomplan::Plan &plan, loom::Id id_base, loom::Dictionary &dictionary)
{
    std::vector<int> resources;

    loom::Id resource_ncpus = dictionary.find_or_create("loom/resource/cpus");
    auto rr_size = plan.resource_requests_size();
    for (int i = 0; i < rr_size; i++) {
        auto &rr = plan.resource_requests(i);
        assert(rr.resources_size() == 1);
        assert(rr.resources(0).resource_type() == resource_ncpus);
        resources.push_back(rr.resources(0).value());
    }

    auto task_size = plan.tasks_size();
    tasks.reserve(task_size);
    for (int i = 0; i < task_size; i++) {
        const auto& pt = plan.tasks(i);
        auto id = i + id_base;

        std::vector<int> inputs;
        auto inputs_size = pt.input_ids_size();
        for (int j = 0; j < inputs_size; j++) {
            inputs.push_back(id_base + pt.input_ids(j));
        }

        int n_cpus = 0;
        if (pt.resource_request_index() != -1) {
            assert(pt.resource_request_index() >= 0);
            assert(pt.resource_request_index() < (int) resources.size());
            n_cpus = resources[pt.resource_request_index()];
        }
        PlanNode pnode(id, i, read_task_policy(pt.policy()), n_cpus, false,
                       pt.task_type(), pt.config(), std::move(inputs));
        tasks.emplace(std::make_pair(id, std::move(pnode)));
    }

    for (auto& pair : tasks) {
        auto& task = pair.second;
        loom::Id my_id = task.id;
        std::vector<loom::Id> inps(task.get_inputs());
        std::sort(inps.begin(), inps.end());
        loom::Id prev = -1;
        for (loom::Id id : inps) {
            if (prev != id) {
                auto it = tasks.find(id);
                assert(it != tasks.end());
                it->second.nexts.push_back(my_id);
                prev = id;
            }
        }
    }

    for (int i = 0; i < plan.result_ids_size(); i++)
    {                
        auto it = tasks.find(plan.result_ids(i) + id_base);
        assert(it != tasks.end());
        it->second.set_as_result();
    }
}

std::vector<loom::Id> Plan::get_init_tasks() const
{
    std::vector<loom::Id> result;
    foreach_task([&result](const PlanNode &node){
          if (node.get_inputs().empty()) {
            result.push_back(node.id);
       }
    });
    return result;
}
